### DISCUSSION TITLE
New accessability for nested type constructors

### DISCUSSION THREAD

**Discussion by Panthr75** [dotnet/csharplang#7416] [Language Ideas]
Consider the following code:

```cs
public class MyCollection : ICollection<string>
{
    // ...
    
    public Enumerator GetEnumerator()
    {
        return new(this);
    }

    //...

    public struct Enumerator : IEnumerator<string>
    {
        internal Enumerator(MyCollection collection)
        {
            // setup logic here
        }

        // ...
    }
}
```

We are forced to make the constructor of Enumerator internal, but that makes the constructor visible to the entire assembly when it only needs to be visible to the `MyCollection` class.

It would be nice to have an accessability modifier to reflect this, or an attribute.

**Comment by Joe4evr** [comment]
~~This is what the new `file` accessibility modifier is for.~~

My bad, looks like that only works on type declarations, not members/constructors.

**Reply by Panthr75** [reply]
Yet putting file is invalid for the constructor:

https://sharplab.io/#v2:CYLg1APgAgTABAQQLACgDeq5blAzD+AIU2wxWwrgDMBLAGwFM5CAKBAShMrjO+68oBfAcJSCgA==

**Comment by DavidArno** [comment]
It looks like you are trying to use nested types create a version of `internal` that only allows access from types in the same namespace. One solution to this is to not nest the types and to instead put them in the same namespace and use an analyzer like [NsDepCop](https://www.nuget.org/packages/NsDepCop) (others likely exist) to make accessing `Enumerator` from anywhere other than from `MyCollection` a compilation error.

Alternatively, such an analyzer is fairly simple to write, so you could write your own custom one that enforces access rules according to your specific requirements.
